### DISCUSSION TITLE
Missing PUPNP error when trying to use  libupnp-dev

### DISCUSSION THREAD

**Discussion by BobManMasher** [gerbera/gerbera#1189] [Q&A] [answered]
I am just following the Ubuntu cmake command line instructions to compile Gerbera for Debian Buster Armel Architecture.  I couldn't get a UPNP version at the correct level so have made an armel version of libnpupnp1. From the instructions, this is an XOR option with pupnp. As far as I can tell, libupnp-dev has been removed and purged from the system but I am getting the following error:

> CMake Error at /usr/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
>   Could NOT find pupnp (missing: UPNP_UPNP_LIBRARY UPNP_INCLUDE_DIR)
>   (Required is at least version "1.12.1")
> Call Stack (most recent call first):
>   /usr/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
>   cmake/Findpupnp.cmake:43 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
>   CMakeLists.txt:340 (find_package)

There is no command line option to specify the upnp version.  I notice that the compiled package for npupnp is called libnpupnp1_4.0.14-1~ppa1~buster_armel.deb i.e. there is a 1 on the end of libnpupnp - might this be an issue?

Any ideas on how to resolve this?

**Comment by whyman** [comment, marked as answer]
Looks like we are missing something from the README table... 

You need to pass -DWITH_NPUPNP=YES to use npupnp

**Reply by BobManMasher** [reply]
That worked - Thanks.

I now have a problem with Duktape though the package is installed
```
CMake Error at /usr/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find duktape (missing: DUKTAPE_LIBRARY DUKTAPE_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/Findduktape.cmake:22 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:412 (find_package)
```
Have I missed something else?

**Reply by BobManMasher** [reply]
Further to this, I just used the command line to not use Duktape as it's an optional library and that resolved things for the moment.

**Reply by whyman** [reply]
We expect to be able to find duktape as a shared library if enabled (it is by default)

https://github.com/gerbera/gerbera/blob/ae283931c491a29b1189347449dd534e3013d8cc/scripts/debian/build-deb.sh#L72-L77

**Reply by BobManMasher** [reply]
Odd that - I just reinstalled both libduktape203 & duktape and they are visible in my list of installed libraries but when I put JS=1 I get the same error.

**Reply by whyman** [reply]
You will need duktape-dev for the header files too

**Reply by KarlStraussberger** [reply]
Did you also install the dev parts for the libraries? On my raspbian it's called `duktape-dev`

**Reply by BobManMasher** [reply]
That's the bit I missed duktape-dev. I now have the cmake command completing with just "spdlog is built without SPDLOG_FMT_EXTERNAL." so off to figure that out. Thank you both.

**Reply by whyman** [reply]
That warning is not _necessarily_ a problem, more of a heads up if you get weird linking problems.

**Reply by BobManMasher** [reply]
Actually that's what I thought too but I kept getting errors during the make pointing to SPDLOG. I tried to build a new version  but was missing a file so ended up just purgeing and reinstalling with APT. Today as if by magic, I get to 91% and the following error:

`[ 91%] Building CXX object CMakeFiles/libgerbera.dir/src/web/items.cc.o
/mnt/sata/gerbera/src/database/sql_database.cc:37:10: fatal error: fmt/chrono.h: No such file or directory
 #include <fmt/chrono.h>
          ^~~~~~~~~~~~~~
compilation terminated.
`
Have purged and reinstalled libfmt-dev (v5.2.1+ds-2) and now trying to find this missing file or whether its supported in my version..

**Reply by BobManMasher** [reply]
So I have a chrono.h on my system located here

./usr/include/spdlog/fmt/bundled/chrono.h

but this seems to be a different path to that expected by the cmake script. Can I simply copy the file into the correct location or modify the cmake script to fix the problem?

**Reply by whyman** [reply]
This is just an indication your fmt is too old. I guess we should add a version check to save this kind of issue.

**Reply by BobManMasher** [reply]
OK - what's the minimum version I need as what I currently have is the Debian latest release, I have tried the version 6 backport too which fails in a different place and I am struggling to compile the very latest version of fmt.

**Reply by BobManMasher** [reply]
Well to answer my own question - yesterday, I finally managed to get a compile to complete.  It may have been a resource problem as the kirkwood armel implementation is low on memory and despite a large swap file on SSD there may still have been problems.  The last thing I did before the compile run was  

1) install FMT 6.1.2+ds-2~bpo10+1 from the buster backport
2) install libspdlog-dev  1:1.5.0+ds-4~bpo10+1 from the buster backport (this also resolved the warning message)

Compile completed successfully,  make install installed the files and an amazing Gerbera server is now running.

**Reply by whyman** [reply]
Great to hear @BobManMasher!
